src/test/java/jenkins/branch/naming/ProjectNamingStrategyTest.java

```diff
@@ -0,0 +1,162 @@
+package jenkins.branch.naming;
+
+import com.sun.net.httpserver.HttpServer;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.logging.Logger;
+import jenkins.branch.BranchSource;
+import jenkins.branch.MultiBranchProjectDisplayNamingStrategy;
+import jenkins.branch.MultiBranchProjectDisplayNamingTrait;
+import jenkins.branch.naming.githubapimock.MockGithubInfo;
+import jenkins.branch.naming.githubapimock.MockGithubJenkinsfileContent;
+import jenkins.branch.naming.githubapimock.MockGithubOrg;
+import jenkins.branch.naming.githubapimock.MockGithubRateLimit;
+import jenkins.branch.naming.githubapimock.MockGithubRepository;
+import jenkins.branch.naming.githubapimock.MockGithubUser;
+import jenkins.branch.naming.githubapimock.MockPullRequest;
+import jenkins.branch.naming.githubapimock.ResourceHandler;
+import jenkins.model.Jenkins;
+import jenkins.scm.api.SCMSource;
+import org.jenkinsci.plugins.github_branch_source.GitHubSCMSource;
+import org.jenkinsci.plugins.github_branch_source.OriginPullRequestDiscoveryTrait;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject;
+import org.junit.*;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * This class tests that project naming strategies (see {@link MultiBranchProjectDisplayNamingStrategy}) applied to
+ * multi-branch jobs via {@link MultiBranchProjectDisplayNamingTrait} are used to set the display name of change
+ * request jobs inside the multi-branch project.
+ *
+ * It does this by using the github-branch-source plugin and stubbing the GitHub REST API to create a fake change
+ * request. Then it checks the job's display name against what
+ * {@link MultiBranchProjectDisplayNamingStrategy#generateName(String, String)} generates for that change request.
+ *
+ * Another option for testing this (that wouldn't have us rely on another plugin) would be to implement our own
+ * fake {@link SCMSource} but that's probably even more complex. Since the github-branch-source plugin is one of the
+ * most used (and fingers-crossed, stable) in the Jenkins ecosystem, and mocking a REST API is trivial, the tradeoff is
+ * acceptable.
+ */
+public class ProjectNamingStrategyTest {
+    private static final Logger LOGGER = Logger.getLogger(ProjectNamingStrategyTest.class.getName());
+
+    @ClassRule
+    public static final JenkinsRule r = new JenkinsRule();
+
+    public static String stubBaseUrl ;
+    public static HttpServer githubApiStub;
+
+    @BeforeClass
+    public static void setupGithubApiStub() throws Exception {
```

Comment:
I wonder how maintainable this would be, there have been changes in the GH API that meant the branch source changes API calls, rather than implementing a few `@TestExtensions`  and/or [Mock SCM implementation](https://github.com/jenkinsci/scm-api-plugin/tree/master/src/test/java/jenkins/scm/impl/mock)

Reply:
Yeah, I tried with mock SCM implementations but ended up in a pretty deep rabbit hole, probably because I'm not aware of how it should be implemented and that's a lot of investment to do to write a single test :sweat_smile: 

Is there any way to re-use an existing mock SCM implementation? Would be happy to re-write this test to use a mock SCM instance but a little more help would go a long way :sweat_smile:

Reply:
[`EnvironmentTest.given_multibranch_when_buildingAChangeRequest_then_environmentContainsBranchName()`](https://github.com/jenkinsci/branch-api-plugin/blob/da3c2aaf0c1160a5aea0146a3e7ec03906d42269/src/test/java/integration/EnvironmentTest.java#L86-L102)  may give you some pointers.

Reply:
I'll have a look, thanks for the pointer.